datasets/seq_cifar100_label_drift.py

```diff
@@ -0,0 +1,228 @@
+
+    def prepare_normal_data(self):
+        pass
+
+
+class TestCIFAR100LabelDrift(MammothDataset, CIFAR100):
+    """Workaround to avoid printing the already downloaded messages."""
+
+    def __init__(self, root, transform, superclass_mapping, metaclass_mapping):
+        self.root = root
+        super().__init__(
+            root,
+            train=False,
+            transform=transform,
+            target_transform=None,
+            download=not self._check_integrity(),
+        )
+        self.superclass_mapping = superclass_mapping
+        self.metaclass_mapping = metaclass_mapping
+        self.classes = list(range(100))
+
+    def __getitem__(self, index: int) -> Tuple[Image.Image, int]:
+        """
+        Args: index (int): Index
+        Returns: tuple: (image, target) where target is index of the target class.
+        """
+        img, target = self.data[index], self.targets[index]
+
+        # doing this so that it is consistent with all other datasets
+        # to return a PIL Image
+        img = Image.fromarray(img)
+        img = self.transform(img)
+        target = self.superclass_mapping[target]
+        target = self.metaclass_mapping[target]
+
+        return img, target
+
+    def select_classes(self, classes_list: list[int]):
+        if len(classes_list) == 0:
+            self.data = np.array([])
+            self.targets = np.array([])
+            self.classes = []
+            return
+
+        mask = np.zeros_like(np.array(self.targets))
+        for label in classes_list:
+            mask = np.logical_or(mask, np.array(self.targets) == label)
+        self.data = self.data[mask]
+        self.targets = np.array(self.targets)[mask]
+
+        self.classes = classes_list
+
+    def apply_drift(self, classes: list):
+        if len(set(self.classes).union(classes)) == 0:
+            return
+        self.drifted_classes.extend(classes)
+
+        # switch meta-class in target_transform dictionary
+        for c in classes:
+            s = self.superclass_mapping[c]
+            self.metaclass_mapping[s] = 0 if self.metaclass_mapping[s] == 1 else 1
+
+    def prepare_normal_data(self):
+        pass
+
+
+class SequentialCIFAR100LabelDrift(ContinualDataset):
+
+    NAME = "seq-cifar100-label-drift"
+    SETTING = "class-il"
+    N_CLASSES_PER_TASK = 5
+    N_TASKS = 20
+
+    TRANSFORM = transforms.Compose(
+        [
+            transforms.RandomCrop(32, padding=4),
+            transforms.RandomHorizontalFlip(),
+            transforms.ToTensor(),
+        ]
+    )
+
+    NO_AUG_TRANSFORM = transforms.Compose([transforms.ToTensor()])
+
+    def get_dataset(self, train=True):
+        """returns native version of represented dataset"""
+
+        if train:
+            return TrainCIFAR100LabelDrift(
+                base_path() + "CIFAR100LabelDrift",
+                transform=self.TRANSFORM,
+                superclass_mapping=superclass_target_mapping,
```

Comment:
I have not catched it right away, but this is wrong. When you pass mutable variable as argument in Python it does not copy that variable but uses the reference instead. It means, that when you pass for example list or dictionary, and modify it inside function, the changes you made are visible in original variable as well (the one you passed to the function). The same goes for classes. So here if you pass global variable to class and change it inside class, the changes will be visible in the global variable as well. This mechanism prevents Python from costly coping of arguments (for example copying of list with 100000 elements when you call some function), but it can lead to some bugs in your code. Need to be careful with it. If you have not heard about mutable vs immutable variables in Python read about them.

So here if you have passed the same variable to train and test dataset, and then you call apply drift, the mapping is changed twice (0 -> 1 -> 0 or 1 -> 0 -> 1). The solution is to use copy.deepcopy on constructor argument.

Reply:
That was a good catch. I know about mutable data types but I did not think of it in this case. Thanks for pointing it out!